stubs/tensorflow/tensorflow/train/__init__.pyi

```diff
@@ -0,0 +1,100 @@
+from _typeshed import Incomplete
+from collections.abc import Callable
+from typing import Any, TypeVar, final
+from typing_extensions import Self
+
+import numpy as np
+import tensorflow as tf
+from google.protobuf.pyext.cpp_message import GeneratedProtocolMessageType
+from tensorflow.python.trackable.base import Trackable
+
+class CheckpointOptions:
+    experimental_io_device: None | str
+    experimental_enable_async_checkpoint: bool
+    # experimental_write_callbacks: None | list[Callable[[str], Any] | Callable[[], Any]]
+    enable_async: bool
+    # experimental_skip_slot_variables: bool
+    # experimental_sharding_callback: tf.train.experimental.ShardingCallback | None = None
+
+    def __init__(
+        self,
+        experimental_io_device: None | str = None,
+        experimental_enable_async_checkpoint: bool = False,
+        # experimental_write_callbacks: None | list[Callable[[str], Any] | Callable[[], Any]] = None,
+        enable_async: bool = False,
+        # experimental_skip_slot_variables: bool = False,
+        # experimental_sharding_callback: tf.train.experimental.ShardingCallback | None = None,
+    ) -> None: ...
+    # def __copy__(self) -> Self: ...
+
+@final
+class Example(GeneratedProtocolMessageType):
```

Comment:
These types are protobuf generated so while there is tf.train.Example, it's not defined here. It's actually defined [here](https://github.com/python/typeshed/blob/main/stubs/tensorflow/tensorflow/core/example/example_pb2.pyi). You should re-export it in this file by doing,

```from tensorflow.core.example.example_pb2 import Example as Example```

Same for rest. mypy-protobuf is responsible for generating protobuf stubs and we have a script here to [build them](https://github.com/python/typeshed/blob/main/scripts/sync_tensorflow_protobuf_stubs.sh). You should not need `GeneratedProtocolMessageType` and can drop cpp_message.pyi

Reply:
@hmc-cs-mdrissi I've tried to re-export the protobuf generated types, but that did not fix the metaclass issues.

Reply:
Hmm looks like we are hitting this [issue](https://github.com/nipunn1313/mypy-protobuf/blame/7f4a558c00faf8fac0cd6d7a6d1332d1643cc08c/stubtest_allowlist.txt#L106). Well if mypy-protobuf produces inconsistent metaclasses I do not thing protobuf stubs should be hand modified. So we should disable those stubtest errors.

There is an allowlist file [here](https://github.com/python/typeshed/blob/main/stubs/tensorflow/%40tests/stubtest_allowlist.txt). You can add `tensorflow.train.BytesList`/along with rest of protobuf types to allowlist and leave a comment that stubtest does not pass for protobuf generated stubs.

Reply:
Done in https://github.com/python/typeshed/pull/11327/commits/794e76935ddf3b5bb6957b99095f432e0706aca4